text/0000-cargo-alternative-registry-auth.md

```diff
@@ -0,0 +1,120 @@
+
+# Guide-level explanation
+Alternative registry operators can set a new key `auth-required = true` in the registry's `config.json` file, which will cause Cargo to include the Authorization token for all API requests, crate downloads, and index updates (if over HTTP).
+
+```json
+{
+    "dl": "https://example.com/index/api/v1/crates",
+    "api": "https://example.com/",
+    "auth-required": true
+}
+```
+
+If the index is hosted via HTTP using [RFC2789](https://github.com/rust-lang/rfcs/pull/2789) and Cargo receives an `HTTP 401` error when fetching `config.json`, Cargo will automatically re-try the request with the Authorization token included.
+
+
+# Reference-level explanation
+A new optional key, `auth-required`, will be allowed in the [`config.json`](https://doc.rust-lang.org/cargo/reference/registries.html#index-format) file stored in the registry index. When this key is set to `true`, the authorization token will be sent with any HTTP requests made to the registry API, crate downloads, and index (if using http). If a token is not available when Cargo is attempting to make a request, the user would be prompted to run `cargo login --registry NAME` to save a token.
+
+The authorization token would be sent as an HTTP header, exactly how it is currently sent for operations such as `publish` or `yank`:
+```
+Authorization: <token>
+```
+
+This RFC does not specify or change the format of the Authorization Token. For the purposes of this RFC, tokens are opaque; no particular format or protocol is specified, and third-party registry authentication should not assume support for any particular format. This includes shared-secret tokens, even though crates.io and the existing publish support for third-party registries currently supports such bearer tokens. Future RFCs (such as [RFC2789](https://github.com/rust-lang/rfcs/pull/3231)) may update the format and protocol used for tokens.
+
+## Interaction with HTTP registries
+The approved (but currently unimplemented) [RFC2789](https://github.com/rust-lang/rfcs/pull/2789) enables Cargo to fetch the index over HTTP. When fetching `config.json` from an HTTP index, if Cargo receives an `HTTP 401` response, the request will be re-attempted with the Authorization header included. If no authorization token is available, Cargo will suggest that the user run `cargo login` to add one. The `HTTP 401` response from the registry server may also include an `X-Cargo-Token-Url: ` header to specify where the user should go to get a token. In that case, `cargo` can display a more helpful message such as "please paste the Token found on https://example.com/token-url-from-header below"

```

Comment:
On the name `X-Cargo-Token-Url`:

- Per [IETF RFC 6648](https://tools.ietf.org/html/rfc6648), `x-` prefixes are deprecated.
- The convention in mixed-case header names is to spellings like “URL” rather than “Url”. (In HTTP/2, it’s all lowercase, and increasingly people are using all lowercase even in HTTP/1, but for some reason even new headers still seem to be consistently defined in mixed case.)
- Also the colon-space present in the text at present is inappropriate, not being part of the header field name.

I would suggest `Cargo-Token-URL` as the header field name to begin with, *if* that’s the semantics stayed with. I agree with @johnterickson that it might be better to just have an arbitrary error message rather than assuming it’ll be in the form of a URL.

I also then feel like noting that there’s kind of already a place for this: the `WWW-Authenticate` header, which is already required on 401 responses ([RFC 7235, §4.1](https://datatracker.ietf.org/doc/html/rfc7235#section-4.1)). I think it would not be unreasonable to define a new challenge scheme `Cargo`, with a parameter `url` or `message` or some such thing—perhaps even allow either, with `message` being more flexible but `url` potentially allowing greater tooling assistance. One could argue this is a mild abuse of the system since this scheme wouldn’t ever be used in the `Authorization` request header but only the `WWW-Authenticate` response header, but I think it’s pretty reasonable. All up, a real example could look like this:

```http
HTTP/1.1 401 Unauthorized
WWW-Authenticate: Basic realm="example.com", Cargo url="https://example.com/token-url-from-header"
```

In the end, I don’t think there’s a great deal to choose between the two approaches (custom header or custom challenge scheme). But no `X-`, please.

Reply:
- ****

Reply:
One thing I don't understand is why Authorization *isn't* being used: a cargo token is exactly what a Bearer token is - RFC 6750 is specifically general and not coupled to OAuth.